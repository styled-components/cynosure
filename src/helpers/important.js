// @flow
import type { Styles } from '../types/style'
import PolishedError from '../internalHelpers/_errors'

/**
 * Helper for targeting rules in a style block generated by polished modules that need !important-level specificity. Can optionally specify a rule (or rules) to target specific rules.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...important(cover())
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${important(cover())}
 * `
 *
 * // CSS as JS Output
 *
 * div: {
 *   'position': 'absolute !important',
 *   'top': '0 !important',
 *   'right: '0 !important',
 *   'bottom': '0 !important',
 *   'left: '0 !important'
 * }
 */
export default function important(
  styleBlock: Styles,
  rules?: Array<string> | string,
): Styles {
  if (typeof styleBlock !== 'object' || styleBlock === null) {
    throw new PolishedError(75, typeof styleBlock)
  }

  Object.keys(styleBlock).forEach((key) => {
    if (typeof styleBlock[key] === 'object' && styleBlock[key] !== null) {
      // eslint-disable-next-line no-param-reassign
      styleBlock[key] = important(styleBlock[key], rules)
    } else if (
      !rules
      || (rules && (rules === key || rules.indexOf(key) >= 0))
    ) {
      // eslint-disable-next-line no-param-reassign
      styleBlock[key] = `${styleBlock[key]} !important`
    }
  })

  return styleBlock
}

// Errors
