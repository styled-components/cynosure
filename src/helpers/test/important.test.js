// @flow
import important from '../important'
import cover from '../../mixins/cover'

describe('important', () => {
  it('should add !important to a single rule in a flat style block', () => {
    expect(important({ color: 'red' })).toEqual({
      color: 'red !important',
    })
  })

  it('should add !important to a single rule in a flat style block when the value is a number', () => {
    expect(important({ fontSize: 12 })).toEqual({
      fontSize: '12 !important',
    })
  })

  it('should add !important to every rule in a flat style block', () => {
    expect(
      important({
        color: 'red',
        background: 'blue',
      }),
    ).toEqual({
      background: 'blue !important',
      color: 'red !important',
    })
  })

  it('should add !important to a target rule when passed as a string', () => {
    expect(
      important(
        {
          color: 'red',
          background: 'blue',
        },
        'color',
      ),
    ).toEqual({
      background: 'blue',
      color: 'red !important',
    })
  })

  it('should add !important to a target rule when passed as a single item array', () => {
    expect(
      important(
        {
          color: 'red',
          background: 'blue',
        },
        ['color'],
      ),
    ).toEqual({
      background: 'blue',
      color: 'red !important',
    })
  })

  it('should add !important to a target rule when passed as an array', () => {
    expect(
      important(
        {
          color: 'red',
          background: 'blue',
          height: '100px',
        },
        ['color', 'height'],
      ),
    ).toEqual({
      background: 'blue',
      color: 'red !important',
      height: '100px !important',
    })
  })

  it('should add !important to a mixture of unnested and nested target rules when passed as an array', () => {
    expect(
      important(
        {
          background: 'blue',
          height: '100px',
          div: {
            color: 'red',
          },
        },
        ['color', 'height'],
      ),
    ).toEqual({
      background: 'blue',
      div: {
        color: 'red !important',
      },
      height: '100px !important',
    })
  })

  it('should add !important to all rules in a polished module', () => {
    expect(important(cover())).toEqual({
      bottom: '0 !important',
      left: '0 !important',
      position: 'absolute !important',
      right: '0 !important',
      top: '0 !important',
    })
  })

  it('should add !important to a specific rule in a polished module', () => {
    expect(important(cover(), 'position')).toEqual({
      bottom: 0,
      left: 0,
      position: 'absolute !important',
      right: 0,
      top: 0,
    })
  })

  it('should return original styles when no properties are found', () => {
    expect(
      important(
        {
          color: 'red',
          background: 'blue',
          height: '100px',
        },
        ['width', 'fontSize'],
      ),
    ).toEqual({
      background: 'blue',
      color: 'red',
      height: '100px',
    })
  })

  it('should throw an error when passed a non-object', () => {
    expect(() => {
      // $FlowFixMe
      important('test')
    }).toThrow('important requires a valid style object, got a string instead.')
  })
})
