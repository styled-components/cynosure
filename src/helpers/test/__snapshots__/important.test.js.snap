// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`important should add !important to a mixture of unnested and nested target rules when passed as an array 1`] = `
Object {
  "background": "blue",
  "div": Object {
    "color": "red !important",
  },
  "height": "100px !important",
}
`;

exports[`important should add !important to a single rule in a flat style block 1`] = `
Object {
  "color": "red !important",
}
`;

exports[`important should add !important to a single rule in a flat style block when the value is a number 1`] = `
Object {
  "fontSize": "12 !important",
}
`;

exports[`important should add !important to a specific rule in a polished module 1`] = `
Object {
  "bottom": 0,
  "left": 0,
  "position": "absolute !important",
  "right": 0,
  "top": 0,
}
`;

exports[`important should add !important to a target rule when passed as a single item array 1`] = `
Object {
  "background": "blue",
  "color": "red !important",
}
`;

exports[`important should add !important to a target rule when passed as a string 1`] = `
Object {
  "background": "blue",
  "color": "red !important",
}
`;

exports[`important should add !important to a target rule when passed as an array 1`] = `
Object {
  "background": "blue",
  "color": "red !important",
  "height": "100px !important",
}
`;

exports[`important should add !important to all rules in a polished module 1`] = `
Object {
  "bottom": "0 !important",
  "left": "0 !important",
  "position": "absolute !important",
  "right": "0 !important",
  "top": "0 !important",
}
`;

exports[`important should add !important to every rule in a flat style block 1`] = `
Object {
  "background": "blue !important",
  "color": "red !important",
}
`;

exports[`important should return original styles when no properties are found 1`] = `
Object {
  "background": "blue",
  "color": "red",
  "height": "100px",
}
`;
